src/shm_manager.cc

```diff
@@ -0,0 +1,189 @@
+#include <unistd.h>
+#include <string>
+
+
+namespace triton { namespace backend { namespace python {
+
+SharedMemory::SharedMemory(const std::string& shm_key)
+{
+  shm_fd_ =
+      shm_open(shm_key.c_str(), O_RDWR | O_CREAT | O_TRUNC, S_IRUSR | S_IWUSR);
+  if (shm_fd_ == -1) {
+    TRITONSERVER_Error* err = TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INTERNAL,
+        ("unable to get shared memory descriptor for shared-memory key '" +
+         shm_key + "'")
+            .c_str());
+    throw BackendModelException(err);
+  }
+
+  size_t default_byte_size = 1024 * 1024 * 128;  // 128MB
+  int res = ftruncate(shm_fd_, default_byte_size);
+  if (res == -1) {
+    TRITONSERVER_Error* err = TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INTERNAL,
+        ("unable to initialize shared-memory key '" + shm_key +
+         "' to requested size: " + std::to_string(default_byte_size) + " bytes")
+            .c_str());
+    throw BackendModelException(err);
+  }
+
+  void* map_addr = mmap(
+      NULL, default_byte_size, PROT_READ | PROT_WRITE, MAP_SHARED, shm_fd_, 0);
+
+  if (map_addr == MAP_FAILED) {
+    auto err = TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INTERNAL,
+        ("unable to process address space or shared-memory descriptor: " +
+         std::to_string(shm_fd_))
+            .c_str());
+    throw BackendModelException(err);
+  }
+  shm_addr_ = (char*)map_addr;
+
+  capacity_ = (size_t*)shm_addr_;
+  *capacity_ = default_byte_size;
+  current_capacity_ = *capacity_;
+
+  // Set offset address
+  offset_ = (off_t*)((char*)shm_addr_ + sizeof(size_t));
+
+  *offset_ = 0;
+  *offset_ += sizeof(off_t);
+  *offset_ += sizeof(size_t);
+
+  shm_key_ = shm_key;
+}
+
+SharedMemory::~SharedMemory() noexcept(false)
+{
+  // TODO: Deallocate the first address
+  for (auto& pair : old_shm_addresses_) {
+    int status = munmap(pair.second, pair.first);
+    if (status == -1) {
+      TRITONSERVER_Error* err = TRITONSERVER_ErrorNew(
+          TRITONSERVER_ERROR_INTERNAL, "unable to munmap shared memory region");
+      throw BackendModelException(err);
+    }
+  }
+
+  // Close fd
+  if (close(shm_fd_) == -1) {
+    TRITONSERVER_Error* err = TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INTERNAL,
+        ("unable to close shared-memory descriptor: " + std::to_string(shm_fd_))
+            .c_str());
+    throw BackendModelException(err);
+  }
+
+  // Unlink shared memory
+  int error = shm_unlink(shm_key_.c_str());
+  if (error == -1) {
+    TRITONSERVER_Error* err = TRITONSERVER_ErrorNew(
+        TRITONSERVER_ERROR_INTERNAL,
+        ("unable to unlink shared memory for key '" + shm_key_ + "'").c_str());
+    throw BackendModelException(err);
+  }
+}
+
+TRITONSERVER_Error*
+SharedMemory::Map(char** shm_addr, size_t byte_size, off_t& offset)
+{
+  while (*offset_ + byte_size >= *capacity_) {
+    // Increase the shared memory pool size by one page size.
+    *capacity_ = *offset_ + byte_size + PAGE_SIZE;
```

Comment:
Why do we only create the capacity by one PAGE_SIZE instead of that fixed size we spoke about earlier? Wil you address this in a future PR?

Reply:
I actually think 128MB is a bit large as the default value.  Because this shared memory region is per model instance which can be a lot if the user has a lot of models (e.g. 10 models will be 1280MB). But I'll fix the growth in this PR along with a paramter to make it controllable.

Reply:
I added the parameter.